### ISSUE TITLE
[Security] Injection detection regex patterns are brittle against paraphrasing and multilingual attacks

### ISSUE THREAD

**Issue by troykelly** [troykelly/openclaw-projects#1256]
## Problem

The 10 injection detection patterns in `packages/openclaw-plugin/src/utils/injection-protection.ts` are English-only regex patterns matching specific phrasings (e.g. "ignore previous instructions", "you are now an unrestricted AI").

Limitations:
- **Paraphrasing**: "Please don't follow your earlier directives" bypasses the patterns
- **Other languages**: Injection attempts in non-English languages are completely invisible
- **Token splitting**: Creative whitespace, Unicode homoglyphs, or word-splitting tricks bypass regex
- **Indirect phrasing**: Socratic-style prompts ("What would happen if you ignored...") won't match

## Context

The current implementation correctly treats detection as monitoring-only (no blocking), which is the right approach. However, the patterns should be acknowledged as a low-confidence canary, not a reliable detection layer.

## Source Files

- **Detection logic**: `packages/openclaw-plugin/src/utils/injection-protection.ts` — `detectInjectionPatterns()` function and `INJECTION_PATTERNS` array
- **Integration points** (where detection is called):
  - `packages/openclaw-plugin/src/register-openclaw.ts` — inline handlers for `message_search`, `thread_get`, `thread_list`
  - `packages/openclaw-plugin/src/tools/message-search.ts` — standalone message_search tool
  - `packages/openclaw-plugin/src/tools/threads.ts` — standalone thread_get/thread_list tools
  - `packages/openclaw-plugin/src/hooks.ts` — auto-recall memory context
- **Tests**: `packages/openclaw-plugin/tests/utils/injection-protection.test.ts`, `packages/openclaw-plugin/tests/injection-protection-integration.test.ts`

## Solution: PromptGuard-2 Classifier

Replace regex-based detection with **Meta's PromptGuard-2** as a classifier-based detection layer. Regex patterns remain as a fast synchronous fallback.

### Model Details
- **Model ID**: `meta-llama/Llama-Prompt-Guard-2-86M`
- **HuggingFace**: https://huggingface.co/meta-llama/Llama-Prompt-Guard-2-86M
- **Architecture**: mDeBERTa-v3-base fine-tuned for prompt injection + jailbreak classification
- **Size**: ~86M parameters, ~500MB on disk
- **Labels**: `BENIGN`, `INJECTION`, `JAILBREAK`
- **Multilingual**: Yes — trained on multilingual data
- **License**: Llama 4 Community License
- **Blog**: https://ai.meta.com/blog/meta-prompt-guard-2/

### Why PromptGuard-2
- Purpose-built for prompt injection and jailbreak detection
- Multilingual — handles non-English attacks that regex cannot
- Handles paraphrasing, indirect phrasing, and obfuscation techniques
- Small model — runs on CPU, no GPU required
- Low latency inference suitable for inline message processing
- Three-class output (benign/injection/jailbreak) with confidence scores

### License Constraints

**The model is gated on HuggingFace** — requires accepting Meta's Llama 4 Community License.

**DO NOT bake model weights into the container image.** The repo and GHCR images are public. Redistributing weights in a public container image bypasses Meta's gated license acceptance flow. Instead:
- Container image contains inference server code only
- Model downloaded at runtime on **first startup only** using `HF_TOKEN`
- Model cached in a **persistent named Docker volume** — survives container rebuilds, restarts, and `docker compose down/up`
- Subsequent startups use cached model (fast startup, no download)
- Include Llama 4 LICENSE and NOTICE file in the container for compliance

Redistribution requirements (if ever distributing weights):
- Must include copy of Llama 4 Community License
- Must prominently display "Built with Llama"
- Must include attribution: "Llama 4 is licensed under the Llama 4 Community License, Copyright (c) Meta Platforms, Inc. All Rights Reserved."

### HuggingFace Auth

- **1Password**: Vault `Claude API Access`, item `HuggingFace - Troy` (ID: `ylxllacyxy3z4g5jr2ysxrfhoy`), field `token`
- **On moltbot**: `/home/moltbot/.hf_token` (mode 0400)
- **Env reference**: `HF_TOKEN_FILE=/home/moltbot/.hf_token` in `~/.moltbot_env`
- **CI**: Not needed — model is not downloaded during CI builds

### Deployment Target

**Host**: moltbot VM (`moltbot.public-servers.sy3.aperim.net`)
- CPU: 6-core Intel i7-8700 @ 3.20GHz
- RAM: 7.6GB total (~4GB available)
- Disk: 45GB free
- OS: Debian 13 (x64)

### Container Spec

**Dockerfile**:
```dockerfile
FROM python:3.12-slim
WORKDIR /app
RUN pip install --no-cache-dir torch --index-url https://download.pytorch.org/whl/cpu \
    && pip install --no-cache-dir transformers fastapi uvicorn
COPY server.py NOTICE LICENSE ./
ENV MODEL_CACHE_DIR=/app/model_cache
ENV MODEL_ID=meta-llama/Llama-Prompt-Guard-2-86M
VOLUME /app/model_cache
EXPOSE 8190
CMD ["uvicorn", "server:app", "--host", "0.0.0.0", "--port", "8190"]
```

**`server.py` startup logic — model persistence is critical:**
1. Check `MODEL_CACHE_DIR` for existing model files (this is a **persistent named Docker volume**)
2. If model files exist → load from cache (fast, no network)
3. If model files missing (first run only) → download from HuggingFace using `HF_TOKEN` env var
4. Save downloaded model to `MODEL_CACHE_DIR` so it persists
5. Load model + tokenizer into memory
6. Expose API endpoints

**The model (~500MB) must only be downloaded ONCE.** The named Docker volume `prompt_guard_model_cache` persists across container rebuilds, image updates, and `docker compose down/up`. Do NOT use anonymous volumes or tmpfs. Do NOT re-download on restart.

**Docker Compose** (add to all compose files, same pattern as Nominatim in #1260):
```yaml
services:
  prompt-guard:
    image: ghcr.io/troykelly/openclaw-projects-prompt-guard:latest
    container_name: openclaw-prompt-guard
    restart: unless-stopped
    environment:
      HF_TOKEN: ${HF_TOKEN:?HF_TOKEN is required for model download}
      MODEL_CACHE_DIR: /app/model_cache
    volumes:
      - prompt_guard_model_cache:/app/model_cache
    ports:
      - "[::1]:8190:8190"
      - "127.0.0.1:8190:8190"
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:8190/health"]
      interval: 30s
      timeout: 10s
      retries: 3
      start_period: 120s
    deploy:
      resources:
        limits:
          cpus: "2"
          memory: 1536M
        reservations:
          cpus: "0.5"
          memory: 768M

volumes:
  prompt_guard_model_cache:
```

Note: `start_period: 120s` accounts for the first-run model download time.

### API Contract

**Service endpoint**: `http://127.0.0.1:8190` (localhost only)

| Endpoint | Method | Request | Response |
|----------|--------|---------|----------|
| `/classify` | POST | `{text: string}` | `{injection: bool, jailbreak: bool, label: string, scores: {benign: float, injection: float, jailbreak: float}}` |
| `/classify/batch` | POST | `{texts: string[]}` | Array of classify results |
| `/health` | GET | — | `{ok: bool, model: string, ready: bool}` |

The `/health` endpoint should return `ready: false` while the model is still downloading/loading.

### Plugin Integration

1. Add `promptGuardUrl` as optional config field in `packages/openclaw-plugin/src/config.ts` (same pattern as `nominatimUrl`)
2. In `detectInjectionPatterns()`, if `promptGuardUrl` is configured:
   - Call `POST {promptGuardUrl}/classify` with the message text
   - Include classifier result in `InjectionDetectionResult` alongside regex matches
   - If classifier service unavailable or not ready, fall back to regex-only (graceful degradation)
3. Regex patterns remain as fast synchronous pre-filter and fallback
4. Log classifier results at same integration points where regex detection is logged
5. Detection remains **monitoring-only** — no blocking based on classifier output
6. Add timeout (e.g. 500ms) to classifier calls to avoid blocking message processing

### Testing

- Unit tests for classifier client (mock HTTP responses)
- Integration test with real container (if available in CI, otherwise skip with `describe.skipIf`)
- False positive tests: benign messages in English, Chinese, Arabic, Spanish should not trigger
- Known injection examples should trigger (English + multilingual)
- Graceful degradation: classifier unavailable falls back to regex
- Latency assertions: classifier response < 100ms for single message on CPU
- Batch endpoint: verify concurrent classification
- Verify `/health` returns `ready: false` during model loading

## References
- Original injection protection PR: #1228
- Original injection protection issue: #1224
- Nonce fix (related): #1255
- Rate limiting for detection logging: #1257
- Truncated snippet detection: #1258

**Comment by aperim-agent** [comment]
## Progress Update: Implementation Complete

### Completed:
- **PromptGuard classifier client** (`packages/openclaw-plugin/src/utils/prompt-guard-client.ts`)
  - `classifyText()`, `classifyBatch()`, `checkHealth()` functions
  - 500ms default timeout, graceful degradation on failure
  
- **Async injection detection** (`detectInjectionPatternsAsync()` in `injection-protection.ts`)
  - Runs regex + classifier in parallel when configured
  - Falls back to regex-only when classifier unavailable
  - Source attribution: `regex`, `classifier`, or `both`
  
- **Plugin config** (`promptGuardUrl` field in config.ts + openclaw.plugin.json)
  - Optional URL field, same pattern as `nominatimUrl`
  - Added to configSchema with `additionalProperties: false` compliance
  - uiHints added for Advanced group

- **Docker container** (`docker/prompt-guard/`)
  - Dockerfile with Python 3.12 + CPU-only PyTorch
  - FastAPI server with /classify, /classify/batch, /health endpoints
  - Model download at runtime via HF_TOKEN, cached in persistent volume
  - Llama 4 LICENSE and NOTICE files included
  
- **Docker Compose** (all 4 compose files updated)
  - prompt-guard service with same hardening pattern as nominatim
  - Named volume `prompt_guard_model_cache`
  - Healthcheck with 120s start_period
  
- **CI** (containers.yml matrix entry added)

- **Callers updated** (threads.ts, message-search.ts, register-openclaw.ts)
  - All detection callsites now use `detectInjectionPatternsAsync()`
  - Classifier results logged alongside regex patterns

### Tests (200 passing):
- 27 prompt-guard-client unit tests (mock HTTP)
- 27 async detection tests (regex+classifier integration)
- 55 existing injection-protection tests (unchanged, still passing)
- 81 config tests (passing with new field)
- 10 integration tests (passing)
